Add CacheFile#fsync() method to ensure cached data are written on disk

tlrx opened the pull request:
This pull request adds a new `CacheFile.fsync()` method to fsync the cache file on disk. This method uses the utility method `IOUtils.fsync(Path, boolean)` that executes a `FileChannel.force()` call under the hood.

tlrx committed:
Add CacheFile#fsync() method to ensure cached data are written on disk

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Snapshot/Restore)

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -398,4 +406,52 @@ private FileChannelReference acquireFileChannelReference() {
+                    try {
+                        // check again if the file is evicted before doing expensive I/O
+                        ensureOpen();
+                        IOUtils.fsync(file, false); // TODO don't forget to fsync parent directory
```
This method is intended to be called periodically by a task that will fsync al cache files of a given shard before fsyncing the parent directory.

original-brownbear commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -305,6 +312,7 @@ protected void doRun() throws Exception {
                             reference.decRef();
                         }
                         gap.onCompletion();
+                        fsynced.set(false);
```
Might be better to set this `false` as early as possible (i.e. directly after the write to the file) to reduce contention statistically speaking? (since `fsync` and `write` will always be blocking each other anyway even if we do use multiple `FileChannel` instances I think).

tlrx replied:
I like that `fsynced` not only indicate that something was written to the file but that ranges are also updated - although it is not a requirement. I'm not sure if it makes a big difference but I can move this earlier.

original-brownbear replied:
Ah that's a good point :)

original-brownbear reviewed (commented):
Just two questions, looks good overall

original-brownbear reviewed (approved):
Either way (regarding the file channels) LGTM, that comment does only matter if the fsync interval we will use is small I guess :)

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -398,4 +406,52 @@ private FileChannelReference acquireFileChannelReference() {
+                    try {
+                        // check again if the file is evicted before doing expensive I/O
+                        ensureOpen();
+                        IOUtils.fsync(file, false); // TODO don't forget to fsync parent directory
```
Yes, I also thought about borrowing the current file channel if we have one but I found that it added more complexity (mostly in incref/decref the file channel another time) to save a file descriptor that would be quickly released. 

I can maybe give it another try by implementing something that use the file channel if there's one or use IOUtils.fsync() otherwise. Deferring the fsync at closing time does not seem the right thing to do as we don't want to fsync everytime a chunk is written but instead control when cache files are fsynced.

henningandersen commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -398,4 +406,52 @@ private FileChannelReference acquireFileChannelReference() {
+     * @throws AlreadyClosedException            if the cache file is evicted
+     * @throws java.nio.file.NoSuchFileException if the cache file does not exist
+     */
+    public List<Tuple<Long, Long>> fsync() throws IOException {
```
Should this return a `SortedSet` instead to signal the expected order verified by the test, as well as uniqueness?

tlrx replied:
This is a good suggestion, thanks. I pushed 7b3590edaa921b328efb5bd5c97162f9795b1c55

henningandersen commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/CacheFileTests.java:
```diff
@@ -171,6 +196,151 @@ public void testConcurrentAccess() throws Exception {
+                }
+
+                cacheFile.startEviction();
+                expectThrows(AlreadyClosedException.class, cacheFile::fsync);
```
This can definitely be handled in a follow-up instead when you start using it. But it might be more appropriate to just return the empty set of ranges in this case, since we essentially having nothing cached for it, rather than have to exception handle in the fsync'er thread?

tlrx replied:
I agree. In case of an evicted file the method will now return an empty collection. See https://github.com/elastic/elasticsearch/pull/64201#discussion_r515932191

henningandersen commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -398,4 +406,52 @@ private FileChannelReference acquireFileChannelReference() {
+                if (fsynced.compareAndSet(false, true)) {
+                    boolean success = false;
+                    try {
+                        // check again if the file is evicted before doing expensive I/O
```
I wonder if this is necessary, since there is no waiting above?

tlrx replied:
I removed it.

henningandersen reviewed (commented):
I added a few comments, looking good otherwise.

tlrx committed:
fsynced no metadata

tlrx committed:
fsynced at creation time

tlrx committed:
sorted set

tlrx committed:
feedback

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -398,4 +406,52 @@ private FileChannelReference acquireFileChannelReference() {
+                final List<Tuple<Long, Long>> completedRanges = tracker.getCompletedRanges();
+                assert completedRanges != null;
+
+                if (fsynced.compareAndSet(false, true)) {
```
I think this is OK, but your comment and your suggestion about returning an empty set of ranges if the cache file got evicted made me change a bit the semantic of this new fsync method. 

I pushed d55278ebf5c6f6ccac3e7af06f3124700215a347 which makes the method return a non-empty set of completed ranges only if the cache file was actually fsynced on disk; otherwise it returns an empty collection. This should address the case of concurrent threads and also simplify the logic to know if new ranges of data were effectively synchronized on disk or not.

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -398,4 +406,52 @@ private FileChannelReference acquireFileChannelReference() {
+                    try {
+                        // check again if the file is evicted before doing expensive I/O
+                        ensureOpen();
+                        IOUtils.fsync(file, false); // TODO don't forget to fsync parent directory
```
Agreed - it is an extra I/O that we don't need. I pushed 6eea9006cc32def270240126785637afad5a3720

tlrx commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/CacheFileTests.java:
```diff
@@ -171,6 +196,151 @@ public void testConcurrentAccess() throws Exception {
+    public void testFSync() throws Exception {
+        try (FSyncTrackingFileSystemProvider fileSystem = setupFSyncCountingFileSystem()) {
+            final CacheFile cacheFile = new CacheFile("test", randomLongBetween(100, 1000), fileSystem.resolve("test"));
+            assertFalse(cacheFile.isFSynced());
```
Right - I pushed 2f8b0e71553feb8c3e6212b4abf0f9ba25c08b22

tlrx commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/CacheFileTests.java:
```diff
@@ -171,6 +196,151 @@ public void testConcurrentAccess() throws Exception {
+                    if (randomBoolean()) {
+                        final long position = i * 10L; // simplify the test by completing small non-contiguous ranges
+                        final Tuple<Long, Long> range = Tuple.tuple(position, position + 1L);
+                        cacheFile.populateAndRead(
```
Sure, I randomized this in the test now.

tlrx commented:
@henningandersen Thanks for your review. I updated the code and changed a bit the semantic of the fsync method. Can you please have another look and tell me what you think? Thanks!

tlrx committed:
Merge branch 'master' into add-cache-file-fsync

tlrx commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into add-cache-file-fsync

tlrx committed:
Merge branch 'master' into add-cache-file-fsync

henningandersen commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -67,6 +70,14 @@ protected void closeInternal() {
+     * considered as fsynced (the initialization value is {@code true}) when it is created; and writing new data to the cache file
+     * will toggle the flag to {@code false}.
+     **/
+    private final AtomicBoolean fsynced = new AtomicBoolean(true);
```
I wonder if calling this `needsFsync` (and reverting the boolean) would be slightly more intuitive?

With the current name, a reader would be inclined to think the boolean is only true when the file is fsync'ed, but this is not true (it is also still true while writing and completing the gap).

tlrx replied:
Thanks, it makes sense. I pushed `d8fdaaa2aab25e78033906f2fa8c956a6e340441`.

henningandersen reviewed (approved):
LGTM, thanks for the extra iteration.

tlrx committed:
needsFsync

tlrx commented:
Thanks Armin and Henning! I'm merging this only on master for now and will backport this change and the futures one about the persistent cache in one go.
